### DISCUSSION TITLE
NextJS: Build once and get API Endpoints URL at runtime

### DISCUSSION THREAD

**Discussion by aniket-kulkarni** [vercel/next.js#39027] [Help]
I am building a new NextJS app. Currently, I am in the stage of deploying the app to multiple environments like dev, qa, staging and prod. All the environments have just one change that it uses different endpoints for API.

Now, I want to create a single dockerfile but use API_ENDPOINTS_URL based on the environment variable passed during the start command.

All the examples that I am see online including NextJS documentation, people are using different build commands for different environments.

```
"build:staging": "cross-env NEXT_PUBLIC_APP_URL=staging.com next build",
"start:staging": "cross-env NEXT_PUBLIC_APP_URL=staging.com next start",
"build:prod": "cross-env NEXT_PUBLIC_APP_URL=production.com next build",
"start:prod": "cross-env NEXT_PUBLIC_APP_URL=production.com next start",
```

Even the example mentioned in the NextJS github example uses 3 different dockerfiles for 3 different environments which is surprising.

https://github.com/vercel/next.js/tree/canary/examples/with-docker-multi-env

Is it possible that I build the app once (one docker image) and send the API_URL as part of environment variable during the start command?

**Comment by piotrcichosz** [comment]
This. 3 different dockerfiles and images are against many standards

**Comment by akkikumar72** [comment]
Facing the same issue, any better way to do this?

**Comment by bluengreen** [comment]
@aniket-kulkarni  I too wasn't really too happy with the way the Next JS example was laid out with multiple docker files. I implemented a solution that used one docker file. In the Next examples you'll notice this line is the main thing thats changing between the envs 

https://github.com/vercel/next.js/blob/canary/examples/with-docker-multi-env/docker/development/Dockerfile#L23

What I did was use a single docker file and set a docker arg that would select the correct env file to pull in for my environment.  Then I created a folder with files that had the env vars I need for each env.  Named like `.env.dev.template`, `.env.production.template` etc 

Like this 

```
ARG STAGE=development
COPY ./environments/.env.${STAGE}.template .env.production.local
```

In docker compose I can call that docker file like normal. When deploying I can set that arg stage to my env.  I'm using AWS Copilot to deploy to ECS Fargate so I just add the arg to the manifest. 

```
image:
  build: 
    dockerfile: Dockerfile
    context: .
    args:
      STAGE: ${COPILOT_ENVIRONMENT_NAME} 
```

You just may have to wiggle those elements around for your specific deploy requirements but hopefully that helps give another idea. Good luck. Let us know what you came up with.

**Comment by jjbowman2** [comment]
I wanted to bump this as I'm facing the same issue at my company. Our CICD expects us to build one docker image and deploy it to all environments. It seems that the current solution builds a different image for each target environment. 

I experimented with loading a different .env.production file at runtime, but it seemed to still use the one provided at build. Is build once, use anywhere possible in the current Next version?

**Comment by osrl** [comment]
Hi, I also want to bump this too.

We're using next.js for over a year now and we liked it a lot. Now we want to use some advanced features like  Output File Tracing and React Server Components. But we're also using `publicRuntimeConfig`. Here is why:

We have three different environments: dev, staging and production. staging and production use the same build with different runtime environment variables. This way, we wanted to make sure staging and production has the exact same build. 

We use Docker and deploy to ECS. So we're using `libc6-compat` and `node:18-alpine`. These are two things that might be updated between two builds. 

I'm not an expert on these topic but I wanted to learn the best practice about this. How do we make sure that two environments has the same build? Should we worry about the changes that might happen during build time?
